SPE-1593/update spiffe auth (#3024)
hashicorp/web-unified-docs: content/vault/v2.x/content/docs/auth/spiffe/index.mdx

Reviewer: "SPIFFE auth is a fit when:" - this section reads a bit confusing to me. Are we specifically referring to the SPIFFE Auth Method, if so we should push SPIRE up to the top. If we are talking about SPIFFE in general, we should remove the reference to Auth.

Author: Yes, this is specific to the auth method side - are you just saying you want the 3rd UL item to be first?

Reviewer: A common point of confusion when *implementing* SPIFFE with Vault is determining which system generates the identity versus which system consumes it. 

Vault cannot issue a SPIFFE identity to a workload before the workload logs in. A SPIFFE control plane (a SPIRE Agent running on the host node, or a service mesh) must attest the workload and issue an SVID first. -- this is confusing. SPIRE issues the SVID, Vault ingests the SVID and returns a token. 

Vault is the consumer, not an issuer of the SVID. The auth/spiffe plugin allows Vault to accept and validate identities generated by an external SPIFFE authority (such as SPIRE), it does not generate SPIFFE identities itself.

Author: Was this meant to be a suggestion? Im not sure what the comment is here

---
layout: docs
page_title: SPIFFE auth methods overview
description: Use the Vault Enterprise SPIFFE auth method to authenticate workloads with X.509 or JWT-based SPIFFE Verifiable Identity Documents (SVIDs) across heterogeneous compute environments.
---

# SPIFFE auth method — overview

@include 'alerts/enterprise-only.mdx'

> The `spiffe` auth method lets workloads authenticate to Vault using SPIFFE

Reviewer 2: A hyperlink to SPIFFE concepts https://spiffe.io/docs/latest/spiffe-about/spiffe-concepts/ before we dive into the SPIFFE world

Verifiable Identity Documents (SVIDs). You can choose either X.509 certificates
or JWTs whose trust comes from a configured trust bundle. Unlike
platform-specific auth methods (Kubernetes, AWS, Azure), SPIFFE auth works the
same way regardless of what the workload is running on: bare metal, virtual
machines (VMs), or multi-cloud Kubernetes clusters.

> 

Author: ```suggestion

If you are not familiar with the SPIFFE framework, review the [SPIFFE concepts](https://spiffe.io/docs/latest/spiffe-about/spiffe-concepts/) documentation.

```

## Use cases for SPIFFE auth with Vault

> SPIFFE auth is a fit when:

Reviewer: I would replace "SPIFFE auth is a fit when:" with "Implementing SPIFFE is a fit when:"

No need to re-order the items.

- Your workloads span heterogeneous compute sources such as bare metal, VMs, or
  multiple clouds with no single platform-native identity issuer covering
  all workloads.
- You need mTLS-based identity (X.509-SVID), not just bearer JWTs. For
  example, if you require Vault access to fit into a broader zero-trust or
  service-mesh identity model rather than a single auth token.
- You already run SPIRE or a service that generates SVIDS for workload identity
  and want Vault to consume identity from that same trust domain instead of
  maintaining a separate auth scheme.
- You need identity that federates across organizational or cloud
  boundaries through a standardized trust-domain model.

For user based interactive/browser-based authentication, review the [JWT/OIDC
auth method](/vault/docs/auth/jwt).

### SPIFFE auth vs. SPIFFE secrets engine

- Use SPIFFE auth (`auth/spiffe`) when you have SPIRE (or another SPIFFE

>   issuer) running and want to log into Vault using a SVID.

Reviewer: replace "want to log into" with "want to authenticate to"

- Use the SPIFFE secrets engine (`secrets/spiffe`) when you are already
  logged into Vault (via Kubernetes, AppRole, AWS, etc.) and want Vault to

>   issue you a SVID. See the SPIFFE secrets engine docs for details.

Reviewer: Use the SPIFFE secrets engine (`secrets/spiffe`) when you want to generate a SPIFFE ID and SVID with an existing Vault auth method (via Kubernetes, AppRole, AWS, etc.). See the SPIFFE secrets engine docs for details.

## Architecture

A common point of confusion when integrating SPIFFE with Vault is
determining which system generates the identity versus which system

> consumes it.

Reviewer: A common point of confusion when implementing SPIFFE with Vault is determining which system generates the identity versus which system consumes it.

Vault cannot issue a SPIFFE identity to a workload before the workload logs in.
A SPIFFE control plane (a SPIRE Agent running on the host node, or a service

> mesh) must attest the workload and issue an SVID first.

Reviewer: I would delete this entire paragraph, it is confusing as related to the auth method.

Author: Given there is both a secrets engine and auth method, I think its important to have here for people familiar with Vault.

> Vault is the consumer, not an issuer of the SVID. The `auth/spiffe` plugin allows

Reviewer: You can add replace your text with this, In regard to the SPIFFE Auth Method, Vault is the consumer, not an issuer of the SVID.  The `auth/spiffe` plugin allows Vault to accept and validate identities generated by an external SPIFFE authority
(such as SPIRE), it does not generate SPIFFE identities itself.

Vault must attest the workload via SVID before issuing a Vault token.

|---|---|---|
| JWKS document | X.509 CA certificate(s) and JWT signing key(s) | X.509-SVIDs and JWT-SVIDs |
| PEM-encoded X.509 certificate | X.509 CA certificate only | X.509-SVIDs only |

The SVID you choose maps directly to the `profile` parameter when you configure
the SPIFFE auth method in Vault. Each type has different use cases and trade
offs.

| Profile | Bundle format | SVID types supported | Vault gets the bundle via | Best for |
|---|---|---|---|---|
| `static` (bundle = PEM) | PEM | X.509-SVIDs only | Pasted in directly, no fetch | Air-gapped/no-network-dependency requirement, JWT-SVID not needed, strict change control on trust-anchor updates |
| `static` (bundle = JWKS) | JWKS | X.509-SVIDs and JWT-SVIDs | Pasted in directly, no fetch | Same as above, but both SVID types needed |
| `https_web_pem` | PEM | X.509-SVIDs only | Auto-fetched over standard TLS | Vault can reach the endpoint, JWT-SVID not needed, endpoint has a normal (web-PKI-trusted) certificate |
| `https_web_bundle` | JWKS | X.509-SVIDs and JWT-SVIDs | Auto-fetched over standard TLS | Vault can reach the endpoint, both SVID types needed, endpoint has a normal (web-PKI-trusted) certificate |

> | `https_spiffe_bundle` | JWKS | X.509-SVIDs and JWT-SVIDs | Auto-fetched, endpoint authenticated via its own SPIFFE ID | Same as above, but you don't want to (or can't) issue the Federation endpoint a publicly-trusted TLS cert — keeps endpoint trust inside the SPIFFE fabric itself instead of web PKI |

Reviewer 3: I don't think this has anything to do with whether the remote endpoint has a public vs private TLS cert - https_web_pem and https_web_bundle profile options support [endpoint_root_ca_truststore_pem](https://developer.hashicorp.com/vault/api-docs/auth/spiffe#endpoint_root_ca_truststore_pem-1). 

From what I can tell there are two differences between web_bundle and spiffe_bundle:
* with spiffe_bundle the remote server can specify the cadence at which we re-fetch updated bundles; with web_bundle we just do it hourly
* with spiffe_bundle instead of using endpoint_root_ca_truststore_pem to validate the remote server's TLS cert, we use the provided bundle for the first request and the last retrieved bundle for subsequent requests
* with spiffe_bundle we require that the SPIFFEID of the remote server's TLS cert match https://developer.hashicorp.com/vault/api-docs/auth/spiffe#endpoint_spiffe_id

Author: You have probably forgotten more about spiffe than i have ever learned, can you please leave a suggestion with what you want it to say.

Reviewer 3: Delete `endpoint has a normal (web-PKI-trusted) certificate ` from all three rows that has it.  As to what the "best for" column for https_spiffe_bundle should say: I'm not sure about the phrasing, but this is the standard option for federating with a spire upstream, and so it's what everyone should be using normally.

If Vault is behind a load balancer or reverse proxy, review [load
balancer and proxy considerations](/vault/docs/auth/spiffe/spiffe#load-balancer-and-proxy-considerations)
before you configure anything. Where TLS terminates determines whether
X.509-SVID auth is viable as documented, or whether you need to rely on
JWT-SVIDs instead.

> ## Choose an auth method

Reviewer 2: I don't think you need this section, at least not here. I think it would be interesting to give more opinionated guidance around which Auth Method to choose in the Vault context overall, but not here.

---
layout: docs
page_title: SPIFFE auth methods overview
description: Use the Vault Enterprise SPIFFE auth method to authenticate workloads with X.509 or JWT-based SPIFFE Verifiable Identity Documents (SVIDs) across heterogeneous compute environments.
---

# SPIFFE auth method — overview

@include 'alerts/enterprise-only.mdx'

The `spiffe` auth method lets workloads authenticate to Vault using SPIFFE
Verifiable Identity Documents (SVIDs). You can choose either X.509 certificates
or JWTs whose trust comes from a configured trust bundle. Unlike
platform-specific auth methods (Kubernetes, AWS, Azure), SPIFFE auth works the
same way regardless of what the workload is running on: bare metal, virtual
machines (VMs), or multi-cloud Kubernetes clusters.

If you are not familiar with the SPIFFE framework, review the [SPIFFE concepts](https://spiffe.io/docs/latest/spiffe-about/spiffe-concepts/) documentation.

## Use cases for SPIFFE auth with Vault

> SPIFFE auth is a fit when:

Author: ```suggestion
Implementing SPIFFE authentication to Vault is a fit when:
```

- Your workloads span heterogeneous compute sources such as bare metal, VMs, or
  multiple clouds with no single platform-native identity issuer covering
  all workloads.
- You need mTLS-based identity (X.509-SVID), not just bearer JWTs. For
  example, if you require Vault access to fit into a broader zero-trust or
  service-mesh identity model rather than a single auth token.
- You already run SPIRE or a service that generates SVIDS for workload identity
  and want Vault to consume identity from that same trust domain instead of
  maintaining a separate auth scheme.
- You need identity that federates across organizational or cloud
  boundaries through a standardized trust-domain model.

For user based interactive/browser-based authentication, review the [JWT/OIDC
auth method](/vault/docs/auth/jwt).

### SPIFFE auth vs. SPIFFE secrets engine

- Use SPIFFE auth (`auth/spiffe`) when you have SPIRE (or another SPIFFE

>   issuer) running and want to log into Vault using a SVID.

Author: ```suggestion
  issuer) running and want to authenticate to Vault using a SVID.
```

- Use the SPIFFE secrets engine (`secrets/spiffe`) when you are already
  logged into Vault (via Kubernetes, AppRole, AWS, etc.) and want Vault to

>   issue you a SVID. See the SPIFFE secrets engine docs for details.

Author: ```suggestion
- Use the SPIFFE secrets engine (secrets/spiffe) when you want to generate a
  SPIFFE ID and SVID with an existing Vault auth method (via Kubernetes, AppRole,
  AWS, etc.). See the SPIFFE secrets engine docs for details.
```

## Architecture

A common point of confusion when integrating SPIFFE with Vault is
determining which system generates the identity versus which system

> consumes it.

Author: ```suggestion
A common point of confusion when implementing SPIFFE with Vault is determining which system generates the identity versus which system consumes it.
```

> Vault cannot issue a SPIFFE identity to a workload before the workload logs in.

Author: ```suggestion
When discussing the SPIFFE auth method, Vault cannot issue a SPIFFE identity to a workload before the workload logs in.
```

|---|---|---|
| JWKS document | X.509 CA certificate(s) and JWT signing key(s) | X.509-SVIDs and JWT-SVIDs |
| PEM-encoded X.509 certificate | X.509 CA certificate only | X.509-SVIDs only |

The SVID you choose maps directly to the `profile` parameter when you configure
the SPIFFE auth method in Vault. Each type has different use cases and trade
offs.

| Profile | Bundle format | SVID types supported | Vault gets the bundle via | Best for |
|---|---|---|---|---|
| `static` (bundle = PEM) | PEM | X.509-SVIDs only | Pasted in directly, no fetch | Air-gapped/no-network-dependency requirement, JWT-SVID not needed, strict change control on trust-anchor updates |
| `static` (bundle = JWKS) | JWKS | X.509-SVIDs and JWT-SVIDs | Pasted in directly, no fetch | Same as above, but both SVID types needed |
| `https_web_pem` | PEM | X.509-SVIDs only | Auto-fetched over standard TLS | Vault can reach the endpoint, JWT-SVID not needed, endpoint has a normal (web-PKI-trusted) certificate |
| `https_web_bundle` | JWKS | X.509-SVIDs and JWT-SVIDs | Auto-fetched over standard TLS | Vault can reach the endpoint, both SVID types needed, endpoint has a normal (web-PKI-trusted) certificate |

> | `https_spiffe_bundle` | JWKS | X.509-SVIDs and JWT-SVIDs | Auto-fetched, endpoint authenticated via its own SPIFFE ID | Same as above, but you don't want to (or can't) issue the Federation endpoint a publicly-trusted TLS cert — keeps endpoint trust inside the SPIFFE fabric itself instead of web PKI |

Author: ```suggestion
| Profile | Bundle format | SVID types supported | Vault gets the bundle via | Best for |
|---|---|---|---|---|
| `static` (bundle = PEM) | PEM | X.509-SVIDs only | Pasted in directly, no fetch | Air-gapped/no-network-dependency requirement, JWT-SVID not needed, strict change control on trust-anchor updates |
| `static` (bundle = JWKS) | JWKS | X.509-SVIDs and JWT-SVIDs | Pasted in directly, no fetch | Same as above, but both SVID types needed |
| `https_web_pem` | PEM | X.509-SVIDs only | Auto-fetched over HTTPS, validated against a CA (system trust store or a custom `endpoint_root_ca_truststore_pem`) | Vault can reach the endpoint, JWT-SVID not needed |
| `https_web_bundle` | JWKS | X.509-SVIDs and JWT-SVIDs | Auto-fetched over HTTPS, validated against a CA (system trust store or a custom `endpoint_root_ca_truststore_pem`) | Federating with a SPIFFE-compliant issuer other than SPIRE, or a SPIRE server specifically exposing an `https_web`-type Federation endpoint |
| `https_spiffe_bundle` | JWKS | X.509-SVIDs and JWT-SVIDs | Auto-fetched; endpoint validated by matching a specific SPIFFE ID (`endpoint_spiffe_id`) rather than a CA chain | The standard choice for federating with a SPIRE upstream — what most setups should use by default |
```

| **Best for** | Workloads in a single k8s cluster | Workloads on a single cloud provider's compute | Workloads whose platform already issues OIDC tokens (k8s, GitHub Actions, GitLab CI, cloud workload identity) | Heterogeneous fleets — bare metal, VMs, multi-cloud, no shared native issuer |
| **Identity source** | Kubernetes API server (projected service account tokens) | Cloud provider IAM / instance metadata | An OIDC issuer's JWKS | A SPIFFE-compliant issuer (SPIRE, Istio, etc.) via node/workload attestation |
| **Credential type** | Bearer JWT | Signed identity document | Bearer JWT | X.509 (mTLS) or JWT |
| **Cross-platform** | No — single cluster | No — single cloud provider | Depends on the issuer | Yes — designed for cross-trust-domain federation |
| **License** | Community Edition | Community Edition | Community Edition | Vault Enterprise |
| **Operational cost** | Low — uses the cluster's built-in issuer | Low — uses the provider's built-in issuer | Low/medium — may need to stand up or trust an IdP | Higher, unless you already run SPIRE/a mesh — server + agents + trust bundle lifecycle |

![Flow chart showing decision points represented in the accompanying auth

> method selection table](/img/diagram-spiffe-auth.png)

Author: ```suggestion
```
